src/backend/Level0/CHIPBackendLevel0.cc

```diff
@@ -359,78 +363,52 @@ CHIPEventLevel0::CHIPEventLevel0(CHIPContextLevel0 *ChipCtx,
 CHIPEventLevel0::CHIPEventLevel0(CHIPContextLevel0 *ChipCtx,
                                  ze_event_handle_t NativeEvent)
     : chipstar::Event((chipstar::Context *)(ChipCtx)), Event_(NativeEvent),
-      EventPoolHandle_(nullptr), Timestamp_(0), EventPoolIndex(0),
-      EventPool(nullptr) {}
+      EventPoolHandle_(nullptr), EventPoolIndex(0), EventPool(nullptr) {}
 
-// Must use this for now - Level Zero hangs when events are host visible +
-// kernel timings are enabled
-void CHIPEventLevel0::recordStream(chipstar::Queue *ChipQueue) {
+void CHIPQueueLevel0::recordEvent(chipstar::Event *ChipEvent) {
   ze_result_t Status;
+  auto ChipEventLz = static_cast<CHIPEventLevel0 *>(ChipEvent);
 
-  {
-    LOCK(EventMtx); // chipstar::Event::EventStatus_
-    if (EventStatus_ == EVENT_STATUS_RECORDED) {
-      logTrace("chipstar::Event {}: EVENT_STATUS_RECORDED ... Resetting event.",
-               (void *)this);
-      ze_result_t Status = zeEventHostReset(Event_);
-      EventStatus_ = EVENT_STATUS_INIT;
-      HostTimestamp_ = 0;
-      DeviceTimestamp_ = 0;
-      CHIPERR_CHECK_LOG_AND_THROW(Status, ZE_RESULT_SUCCESS, hipErrorTbd);
-    }
-  }
+  ChipEventLz->reset();
 
-  auto Dev = (CHIPDeviceLevel0 *)ChipQueue->getDevice();
-  Status = zeDeviceGetGlobalTimestamps(Dev->get(), &HostTimestamp_,
-                                       &DeviceTimestamp_);
+  Status = zeDeviceGetGlobalTimestamps(ChipDevLz_->get(),
+                                       &ChipEventLz->getHostTimestamp(),
+                                       &ChipEventLz->getDeviceTimestamp());
   CHIPERR_CHECK_LOG_AND_THROW(Status, ZE_RESULT_SUCCESS, hipErrorTbd);
 
-  if (ChipQueue == nullptr)
-    CHIPERR_LOG_AND_THROW("Queue passed in is null", hipErrorTbd);
+  ze_command_list_handle_t CommandList = ChipCtxLz_->getCmdListReg();
+
+  auto EventsToWaitOn = getSyncQueuesLastEvents();
+  auto EventToWaitOnHandles = getEventListHandles(EventsToWaitOn);
+
+  // create an Event for making a dependency chain
+  std::shared_ptr<chipstar::Event> TimestampWriteComplete =
+      static_cast<CHIPBackendLevel0 *>(Backend)->createCHIPEvent(
+          this->ChipContext_);
```

Comment:
Casting does not seem to be needed.
```suggestion
  std::shared_ptr<chipstar::Event> TimestampWriteComplete =
      Backend->createCHIPEvent(this->ChipContext_);
```